File: backend/internal/handlers/telemetry.go
package handlers

import (
	"backend/internal/logger"
	"encoding/json"
	"io"
	"net/http"
)

type TelemetryHandler struct {
	logger *logger.Logger
}

type TelemetryEvent struct {
	Type      string                 `json:"type"`
	Data      map[string]interface{} `json:"data"`
	Timestamp string                 `json:"timestamp"`
	SessionID string                 `json:"sessionId"`
	UserID    string                 `json:"userId"`
	UserAgent string                 `json:"userAgent"`
	URL       string                 `json:"url"`
	Referrer  string                 `json:"referrer"`
}

func NewTelemetryHandler(log *logger.Logger) *TelemetryHandler {
	return &TelemetryHandler{
		logger: log,
	}
}

func (h *TelemetryHandler) HandleTelemetry(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		h.logger.Error("telemetry", "Failed to read request body", map[string]interface{}{
			"error": err.Error(),
		})
		http.Error(w, "Failed to read request body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	var event TelemetryEvent
	if err := json.Unmarshal(body, &event); err != nil {
		h.logger.Warn("telemetry", "Failed to parse telemetry event", map[string]interface{}{
			"error": err.Error(),
			"body":  string(body),
		})
		http.Error(w, "Invalid JSON payload", http.StatusBadRequest)
		return
	}

	// Log the telemetry event with appropriate level based on type
	logData := map[string]interface{}{
		"event_type": event.Type,
		"user_id":    event.UserID,
		"session_id": event.SessionID,
		"url":        event.URL,
		"user_agent": event.UserAgent,
	}

	// Merge event data
	for k, v := range event.Data {
		logData[k] = v
	}

	// Log with different messages based on event type
	switch event.Type {
	case "click":
		element, _ := event.Data["element"].(string)
		h.logger.Info("telemetry", "User clicked: "+element, logData)
	case "command":
		command, _ := event.Data["command"].(string)
		h.logger.Info("telemetry", "User executed command: "+command, logData)
	case "navigation":
		to, _ := event.Data["to"].(string)
		h.logger.Info("telemetry", "User navigated to: "+to, logData)
	case "pageview":
		page, _ := event.Data["page"].(string)
		h.logger.Info("telemetry", "User viewed page: "+page, logData)
	case "interaction":
		action, _ := event.Data["action"].(string)
		h.logger.Info("telemetry", "User interaction: "+action, logData)
	default:
		h.logger.Debug("telemetry", "Unknown telemetry event", logData)
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]string{
		"status": "success",
	})
}

File: frontend/src/services/telemetry.ts
type TelemetryEvent = {
  type: "click" | "command" | "navigation" | "pageview" | "interaction";
  data: Record<string, unknown>;
  timestamp?: string;
};

class TelemetryService {
  private endpoint = "/api/telemetry";
  private sessionId: string;
  private userId: string;

  constructor() {
    // Generate or retrieve session ID
    this.sessionId = this.getOrCreateSessionId();
    this.userId = this.getOrCreateUserId();
  }

  private getOrCreateSessionId(): string {
    let sessionId = sessionStorage.getItem("telemetry_session_id");
    if (!sessionId) {
      sessionId = `session_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
      sessionStorage.setItem("telemetry_session_id", sessionId);
    }
    return sessionId;
  }

  private getOrCreateUserId(): string {
    let userId = localStorage.getItem("telemetry_user_id");
    if (!userId) {
      userId = `user_${Date.now()}_${Math.random().toString(36).substring(2, 15)}`;
      localStorage.setItem("telemetry_user_id", userId);
    }
    return userId;
  }

  private async send(event: TelemetryEvent) {
    try {
      const payload = {
        ...event,
        timestamp: new Date().toISOString(),
        sessionId: this.sessionId,
        userId: this.userId,
        userAgent: navigator.userAgent,
        url: window.location.href,
        referrer: document.referrer,
      };

      await fetch(this.endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });
    } catch (error) {
      console.error("Failed to send telemetry:", error);
    }
  }

  trackClick(element: string, data?: Record<string, unknown>) {
    this.send({
      type: "click",
      data: {
        element,
        ...data,
      },
    });
  }

  trackCommand(command: string, success: boolean = true) {
    this.send({
      type: "command",
      data: {
        command,
        success,
      },
    });
  }

  trackNavigation(from: string, to: string) {
    this.send({
      type: "navigation",
      data: {
        from,
        to,
      },
    });
  }

  trackPageView(page: string) {
    this.send({
      type: "pageview",
      data: {
        page,
        title: document.title,
      },
    });
  }

  trackInteraction(action: string, data?: Record<string, unknown>) {
    this.send({
      type: "interaction",
      data: {
        action,
        ...data,
      },
    });
  }
}

// Export singleton instance
export const telemetry = new TelemetryService();
